README.md
# VIVO Tools

Simple python functions for managing data in VIVO

See the test scripts for examples of use.


## Plans for VIVO Tools

### Version 2.0

Version 2.0 will support VIVO 1.6 and VIVO-ISF. We expect to 
complete VIVO Tools 2.0 by September 2015.  

Additionally,
VIVO Tools 2.0 will have modules and a setup.py.  We tried hard
to keep everything very simple -- one file, no install procedure,
one import statement -- but best practice is to organize the
code into smaller modules and use only the modules one needs.

#### Planned modules

Each module will have the input/output functions for managing entities
in the domain

1. Grants -- grants, studies, roles, and related entities
1. People -- contact information, positions, education, overviews, service
1. Courses -- courses, course sections, academic terms, roles
1. Organizations -- departments, universities, publishers and many more
1. Publications -- including all scholarly works, authorships 
1. Foundation -- uri, rdf, assertion level processing, including input/output
to and from the VIVO triple store

#### Design Goals

1. Minimize requirements for additional libraries.  Keep it simple.
1. Consistent use of functions and identifiers
1. Improve separation of assertion level (triples, rdf, uri, ontology) and
domain level (person, grant, pub, org, course, role)
1. Increased emphasis on output functions for visualization and reporting
1. Isolation of UF specific features and dependencies

### Version 3.0

VIVO Tools will move toward an object-oriented representation of domain
entities.  We expect to collaborate with other emerging VIVO apps and tools
projects to provide useful functionality for others.


CHANGELOG.txt
VIVO Tools Change log
    
    1.1     2013-03-02 MC
            Added assert_data_property to generate RDF for asserting that an
            entity have a named data property value
            Added update_data property to generate addtion and subtraction RDF
            based on five case logic to update a VIVO data property value with
            a source data property value
            Added hr_abbrev_to_words to fix position and working titles
    1.2     2013-06-11 MC
            Add get_position to return a position object from a position uri
            Add position capability to get_person
            Add repair_phone_number to return IT standard phone numbers
    1.3     2013-07-23 MC
            Fixed make_datetime_rdf to provide datetime string values with
            zeroed time components for datetime precision asserted as
            yearMonthDay
    1.4     2013-08-07 MC
            Changed delimiter in read_csv to "|" was "!"
    1.41    2013-08-11 MC
            support missing grant start and end dates
    1.42    2013-08-22 MC
            added Skip=True to read.csv to skip rows without the number of
            values in the header.  skip=False will throw an exception on
            first such occurance
    1.43    2013-09-02 MC
            get_person now returns preferred_title.
            get_person now returns person['home_dept_uri'] if exists
    1.44    2013-09-04 MC
            Added version variable
    1.45    2013-09-17 MC
            Added debug parameter to vivo_sparql_query to show baseURL and query
    1.46    2013-11-29 MC
            Remove print statements from update_data_property and
            update_resource_property
            Added get_references to return references to a URI
            Added remove_uri to remove all triples citing the URI
            Added get_webpage to return a webpage object from a webpage uri
            Enhancements to get_publication to return pubmed values
            Removed function _values_from_vivo.  Use get_publication
            Renamed get_pubmed_values_from_pubmed to simply
            get_pubmed_values
    1.47    2013-12-10 MC
            Added additional ttributes to get_pubmed and get_publication
    1.48    2013-12-14 MC
            Add retry code to get_pubmed to handle Entrez non-response
    1.49    2014-01-02 MC
            Numerous upgrades to get functions to support biosketches
    1.50    2014-01-11 MC
            Replace all tests for None with is None and is not None
            VIVO URI always appear in rdf:resource assertions
    1.51    2014-03-25 MC
            find_vivo_uri will return a single URI having the predicate and
            value. Useful for finding uri from other identifier such as uri
            ufid
    1.52    2014-03-28 MC
            Add lang and datatype support to assert_data_property.  New
            function get_value returns lang and datatype. Add newline to
            rdf_header.  Add xml declaration to rdf_header.  Add newline to
            rdf_footer.  Add dict support to update_data_property
    1.53    2014-03-29 MC  
            Add lang and datatype support to get_grant
    1.54    2014-04-06 MC
            Add unicode support for UTF-8 to read_csv
    1.55    2014-05-04 MC
            Restore escape to update data_property
    1.56    2014-05-05 MC
            merge_uri to create add and sub rdf for merging to uris.
            remove_uri updated to handle xml lang and datatype
            translate_predicate added to support remove_uri and merge_uri
    1.57    2014-05-19 MC
            Patch get_publication to handle authorship with no author rank
            Add make_date_dictionary
    1.58    2014-06-03 MC
            Improvements in document_from_pubmed to handle missing data in
            Entrez. Add untag_predicate. Adjust new line in
            assert_resource_property to match new line usage
            assert_data_property. New parameters in update_pubmed to support
            adding pubmed attributes to papers not yet in VIVO. New
            parameters added to get_pubmed_values to support getting values
            from either doi or pmid
    1.59    2014-06-21 MC
            get_types returns a list of types for a specified URI

test_get_role.py
"""
    test_get_role.py -- Given a URI of a role, return a python structure
    representing the attributes of the role

    Version 0.1 MC 2013-12-27
    --  Initial version.
    Version 0.2 MC 2014-01-04
    --  Use JSON dumps

    To Do
    --  Generalize get_role.  It works only for some grant roles
    
"""

__author__ = "Michael Conlon"
__copyright__ = "Copyright 2014, University of Florida"
__license__ = "BSD 3-Clause license"
__version__ = "0.2"

import vivotools as vt
from datetime import datetime
import json

print datetime.now(),"Start"
roles = [
    "http://vivo.ufl.edu/individual/n1665912700",
    "http://vivo.ufl.edu/individual/n206828627",
    "http://vivo.ufl.edu/individual/n1901240698",
    "http://vivo.ufl.edu/individual/n9688",
    "http://vivo.ufl.edu/individual/n1796432519",
    "http://vivo.ufl.edu/individual/n468916789",
    "http://vivo.ufl.edu/individual/n1216606496",
    "http://vivo.ufl.edu/individual/n1613740695",
    "http://vivo.ufl.edu/individual/n1071648987",
    "http://vivo.ufl.edu/individual/n270897"]
for role in roles:
    print "\n",json.dumps(vt.get_role(role),indent=4,sort_keys=True)
print datetime.now(),"Finish"


test_read_csv.py
"""
    test_read_csv.py -- given a CSV files, read it into a python dictionary

    Version 0.1 MC 2013-12-27
    --  Initial version.
"""

__author__      = "Michael Conlon"
__copyright__   = "Copyright 2013, University of Florida"
__license__     = "BSD 3-Clause license"
__version__     = "0.1"

import vivotools as vt
from datetime import datetime

print datetime.now(),"Start"

data = vt.read_csv("data_test_read_csv.csv")
print "data file has ",len(data.keys()),"rows and ",len(data[1]),"columns"
print "The data as one structure:",data
print "Display the data as rows and columns:"

names = data[1].keys()
for name in names:
    print name,
print

for row in sorted(data.keys()):
    line = data[row]
    for val in line.values():
        print val,
    print


print datetime.now(),"Finish"


test_get_grant.py
"""
    test_get_grant.py -- Given a URI of a grant, return a python
    structure representing the attributes of the grant

    Version 0.1 MC 2013-12-27
    --  Initial version
    Version 0.2 MC 2014-01-04
    --  Format output using json.dumps

"""

__author__ = "Michael Conlon"
__copyright__ = "Copyright 2014, University of Florida"
__license__ = "BSD 3-Clause license"
__version__ = "0.2"

import vivotools as vt
from datetime import datetime
import json

print datetime.now(),"Start"
grants = [
    "http://vivo.ufl.edu/individual/n1769500492",
    "http://vivo.ufl.edu/individual/n678985748",
    "http://vivo.ufl.edu/individual/n1725957599",
    "http://vivo.ufl.edu/individual/n859986285",
    "http://vivo.ufl.edu/individual/n899821550",
    "http://vivo.ufl.edu/individual/n2026295266",
    "http://vivo.ufl.edu/individual/n719692692",
    "http://vivo.ufl.edu/individual/n482305201",
    "http://vivo.ufl.edu/individual/n276081150",
    "http://vivo.ufl.edu/individual/n1863481639"
    ]
for grant in grants:
    print "\n", json.dumps(vt.get_grant(grant, get_investigators=False), \
        indent=4, sort_keys=True)
print datetime.now(),"Finish"


demo_date_compare.py
"""
    demo_date_compare.py -- Show comparisons of dates in python

    Version 0.1 MC 2013-12-28
    --  Initial version.
"""

__author__      = "Michael Conlon"
__copyright__   = "Copyright 2013, University of Florida"
__license__     = "BSD 3-Clause license"
__version__     = "0.1"

import dateutil.parser

date1_text = "1968-01-01T00:00:00Z"
date1 = dateutil.parser.parse(date1_text)
#date1 = date1.replace(tzinfo=dateutil.tz.tzlocal())

print "Date 1 Text  ",date1_text
print "Date 1 Format",date1.isoformat()

date2_text = "2013-01-31T05:00:00Z"
date2 = dateutil.parser.parse(date2_text)

print "Date 2 Text  ",date2_text
print "Date 2 Format",date2.isoformat()

print "Text 1 == 2?",date1_text==date2_text
print "Date 1 == 2?",date1==date2

date3_text = "2013-01-31T00:00:00"
date3 = dateutil.parser.parse(date3_text)

print "Date 3 Text  ",date3_text
print "Date 3 Format",date3.isoformat()

print "Text 1 == 3?",date1_text==date3_text
print "Date 1 == 3?",date1==date3


demo_sample_uf_faculty.py
"""
    demo_sample_uf_faculty.py -- Generate a random sample of current UF faculty

    Version 0.1 MC 2013-12-27
    --  Initial version.
"""

__author__      = "Michael Conlon"
__copyright__   = "Copyright 2013, University of Florida"
__license__     = "BSD 3-Clause license"
__version__     = "0.1"

import vivotools as vt
import random
from datetime import datetime

query = """
    SELECT ?uri WHERE
    {
    ?uri a vivo:FacultyMember .
    ?uri a ufVivo:UFCurrentEntity .
    }
"""
print datetime.now(),"Gathering Current UF Faculty from VIVO"
data = vt.vivo_sparql_query(query) 
print datetime.now(),"Current UF Faculty found = ",len(data["results"]["bindings"])
print datetime.now(),"Load data structure with results"
d = []
for item in data["results"]["bindings"]:
    d.append(item["uri"]["value"])
print datetime.now(),"Select random sample"
random.shuffle(d)
print datetime.now(),"Show selected faculty by VIVO URI"
for i in range(100):
    print d[i]
print datetime.now(),"Finished"

